<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_27) on Fri Mar 22 08:30:56 EST 2013 -->
<TITLE>
IntegerAllele
</TITLE>

<META NAME="date" CONTENT="2013-03-22">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="IntegerAllele";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/IntegerAllele.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/jgapcustomised/impl/DefaultConfiguration.html" title="class in org.jgapcustomised.impl"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../org/jgapcustomised/impl/Pool.html" title="class in org.jgapcustomised.impl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/jgapcustomised/impl/IntegerAllele.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="IntegerAllele.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgapcustomised.impl</FONT>
<BR>
Class IntegerAllele</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">org.jgapcustomised.Allele</A>
      <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><B>org.jgapcustomised.impl.IntegerAllele</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.lang.Comparable</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>IntegerAllele</B><DT>extends <A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></DL>
</PRE>

<P>
A Gene implementation that supports a integer values for its allele. Upper and lower bounds may optionally be
 provided to restrict the range of legal values allowed by this Gene instance.
<P>

<P>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#INTEGER_RANGE">INTEGER_RANGE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Represents the constant range of values supported by integers.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#m_activeConfiguration">m_activeConfiguration</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The current active configuration that is in use.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#m_boundsUnitsToIntegerUnits">m_boundsUnitsToIntegerUnits</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Stores the number of integer range units that a single bounds-range unit represents.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#m_lowerBounds">m_lowerBounds</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The lower bounds of values represented by this Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#m_upperBounds">m_upperBounds</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The upper bounds of values represented by this Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.lang.Integer</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#m_value">m_value</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;References the internal integer value (allele) of this Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#PERSISTENT_FIELD_DELIMITER">PERSISTENT_FIELD_DELIMITER</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Represents the delimiter that is used to separate fields in the persistent representation of IntegerGene
 instances.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#IntegerAllele()">IntegerAllele</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new IntegerGene with default settings.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#IntegerAllele(org.jgapcustomised.Configuration)">IntegerAllele</A></B>(<A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A>&nbsp;a_activeConfiguration)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new IntegerGene according to the given active configuration.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#IntegerAllele(org.jgapcustomised.Configuration, int, int)">IntegerAllele</A></B>(<A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A>&nbsp;a_activeConfiguration,
              int&nbsp;a_lowerBounds,
              int&nbsp;a_upperBounds)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new IntegerGene with the given active configuration and the specified lower and upper bounds for
 values represented by this Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#IntegerAllele(int, int)">IntegerAllele</A></B>(int&nbsp;a_lowerBounds,
              int&nbsp;a_upperBounds)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new IntegerGene with the specified lower and upper bounds for values (alleles) of this Gene
 instance.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#calculateBoundsUnitsToIntegerUnitsRatio()">calculateBoundsUnitsToIntegerUnitsRatio</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates and sets the m_boundsUnitsToIntegerUnits field based
 
 on the current lower and upper bounds of this IntegerGene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#cleanup()">cleanup</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Executed by the genetic engine when this Gene instance is no
 
 longer needed and should perform any necessary resource cleanup.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#cloneAllele()">cloneAllele</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#distance(org.jgapcustomised.Allele)">distance</A></B>(<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A>&nbsp;target)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#getPersistentRepresentation()">getPersistentRepresentation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a string representation of this Gene that includes any
 
 information required to reconstruct it at a later time, such as its
 
 value and internal state.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Integer</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#getValue()">getValue</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#hashCode()">hashCode</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the hash code value for this IntegerGene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#intValue()">intValue</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the int value of this Gene, which may be more convenient in
 
 some cases than the more general getAllele() method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#mapValueToWithinBounds()">mapValueToWithinBounds</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Maps the value of this IntegerGene to within the bounds specified by
 
 the m_upperBounds and m_lowerBounds instance variables.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#newAllele(org.jgapcustomised.Configuration)">newAllele</A></B>(<A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A>&nbsp;a_activeConfiguration)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Provides an implementation-independent means for creating new Gene instances.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#setToRandomValue(java.util.Random, boolean)">setToRandomValue</A></B>(java.util.Random&nbsp;a_numberGenerator,
                 boolean&nbsp;onlyPerturbFromCurrentValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value (allele) of this Gene to a random Integer value between
 
 the lower and upper bounds (if any) of this Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#setValue(java.lang.Integer)">setValue</A></B>(java.lang.Integer&nbsp;a_newValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value (allele) of this Gene to the new given value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A></B>(java.lang.String&nbsp;a_representation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value and internal state of this Gene from the string
 
 representation returned by a previous invocation of the
 
 getPersistentRepresentation() method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a string representation of this IntegerGene's value that
 
 may be useful for display purposes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgapcustomised/impl/IntegerAllele.html#toXml()">toXml</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.jgapcustomised.Allele"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class org.jgapcustomised.<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#compareTo(java.lang.Object)">compareTo</A>, <A HREF="../../../org/jgapcustomised/Allele.html#equals(java.lang.Object)">equals</A>, <A HREF="../../../org/jgapcustomised/Allele.html#getGene()">getGene</A>, <A HREF="../../../org/jgapcustomised/Allele.html#getGenes(java.util.Collection)">getGenes</A>, <A HREF="../../../org/jgapcustomised/Allele.html#getInnovationId()">getInnovationId</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, finalize, getClass, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="INTEGER_RANGE"><!-- --></A><H3>
INTEGER_RANGE</H3>
<PRE>
protected static final long <B>INTEGER_RANGE</B></PRE>
<DL>
<DD>Represents the constant range of values supported by integers.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#org.jgapcustomised.impl.IntegerAllele.INTEGER_RANGE">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="PERSISTENT_FIELD_DELIMITER"><!-- --></A><H3>
PERSISTENT_FIELD_DELIMITER</H3>
<PRE>
protected static final java.lang.String <B>PERSISTENT_FIELD_DELIMITER</B></PRE>
<DL>
<DD>Represents the delimiter that is used to separate fields in the persistent representation of IntegerGene
 instances.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#org.jgapcustomised.impl.IntegerAllele.PERSISTENT_FIELD_DELIMITER">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="m_value"><!-- --></A><H3>
m_value</H3>
<PRE>
protected java.lang.Integer <B>m_value</B></PRE>
<DL>
<DD>References the internal integer value (allele) of this Gene.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_upperBounds"><!-- --></A><H3>
m_upperBounds</H3>
<PRE>
protected int <B>m_upperBounds</B></PRE>
<DL>
<DD>The upper bounds of values represented by this Gene. If not explicitly provided by the user, this should be set
 to Integer.MAX_VALUE.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_lowerBounds"><!-- --></A><H3>
m_lowerBounds</H3>
<PRE>
protected int <B>m_lowerBounds</B></PRE>
<DL>
<DD>The lower bounds of values represented by this Gene. If not explicitly provided by the user, this should be set
 to Integer.MIN_VALUE
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_boundsUnitsToIntegerUnits"><!-- --></A><H3>
m_boundsUnitsToIntegerUnits</H3>
<PRE>
protected long <B>m_boundsUnitsToIntegerUnits</B></PRE>
<DL>
<DD>Stores the number of integer range units that a single bounds-range unit represents. For example, if the integer
 range is -2 billion to +2 billion and the bounds range is -1 billion to +1 billion, then each unit in the bounds
 range would map to 2 units in the integer range. The value of this variable would therefore be 2. This mapping
 unit is used to map illegal allele values that are outside of the bounds to legal allele values that are within
 the bounds.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_activeConfiguration"><!-- --></A><H3>
m_activeConfiguration</H3>
<PRE>
protected <A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A> <B>m_activeConfiguration</B></PRE>
<DL>
<DD>The current active configuration that is in use.
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="IntegerAllele()"><!-- --></A><H3>
IntegerAllele</H3>
<PRE>
public <B>IntegerAllele</B>()</PRE>
<DL>
<DD>Constructs a new IntegerGene with default settings. No bounds will be put into effect for values (alleles) of
 this Gene instance, other than the standard range of integer values.
<P>
</DL>
<HR>

<A NAME="IntegerAllele(int, int)"><!-- --></A><H3>
IntegerAllele</H3>
<PRE>
public <B>IntegerAllele</B>(int&nbsp;a_lowerBounds,
                     int&nbsp;a_upperBounds)</PRE>
<DL>
<DD>Constructs a new IntegerGene with the specified lower and upper bounds for values (alleles) of this Gene
 instance.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_lowerBounds</CODE> - The lowest value that this Gene may possess, inclusive.<DD><CODE>a_upperBounds</CODE> - The highest value that this Gene may possess, inclusive.</DL>
</DL>
<HR>

<A NAME="IntegerAllele(org.jgapcustomised.Configuration)"><!-- --></A><H3>
IntegerAllele</H3>
<PRE>
public <B>IntegerAllele</B>(<A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A>&nbsp;a_activeConfiguration)</PRE>
<DL>
<DD>Constructs a new IntegerGene according to the given active configuration.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_activeConfiguration</CODE> - The current active configuration.</DL>
</DL>
<HR>

<A NAME="IntegerAllele(org.jgapcustomised.Configuration, int, int)"><!-- --></A><H3>
IntegerAllele</H3>
<PRE>
public <B>IntegerAllele</B>(<A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A>&nbsp;a_activeConfiguration,
                     int&nbsp;a_lowerBounds,
                     int&nbsp;a_upperBounds)</PRE>
<DL>
<DD>Constructs a new IntegerGene with the given active configuration and the specified lower and upper bounds for
 values represented by this Gene.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_activeConfiguration</CODE> - The current active configuration.<DD><CODE>a_lowerBounds</CODE> - The lowest value that this Gene may represent, inclusive.<DD><CODE>a_upperBounds</CODE> - The highest value that this Gene may represent, inclusive.</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="toXml()"><!-- --></A><H3>
toXml</H3>
<PRE>
public java.lang.String <B>toXml</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>XML representation of object</DL>
</DD>
</DL>
<HR>

<A NAME="distance(org.jgapcustomised.Allele)"><!-- --></A><H3>
distance</H3>
<PRE>
public double <B>distance</B>(<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A>&nbsp;target)</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#distance(org.jgapcustomised.Allele)">distance</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>target</CODE> - 
<DT><B>Returns:</B><DD>compatibility distance based on intValue; always positive<DT><B>See Also:</B><DD><A HREF="../../../org/jgapcustomised/Allele.html#distance(org.jgapcustomised.Allele)"><CODE>Allele.distance(Allele)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="newAllele(org.jgapcustomised.Configuration)"><!-- --></A><H3>
newAllele</H3>
<PRE>
public <A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A> <B>newAllele</B>(<A HREF="../../../org/jgapcustomised/Configuration.html" title="class in org.jgapcustomised">Configuration</A>&nbsp;a_activeConfiguration)</PRE>
<DL>
<DD>Provides an implementation-independent means for creating new Gene instances. The new instance that is created
 and returned should be setup with any implementation-dependent configuration that this Gene instance is setup
 with (aside from the actual value, of course). For example, if this Gene were setup with bounds on its value,
 then the Gene instance returned from this method should also be setup with those same bounds. This is important,
 as the JGAP core will invoke this method on each Gene in the sample Chromosome in order to create each new Gene
 in the same respective gene position for a new Chromosome.
 <p>
 It should be noted that nothing is guaranteed about the actual value of the returned Gene and it should therefore
 be considered to be undefined.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_activeConfiguration</CODE> - The current active configuration.
<DT><B>Returns:</B><DD>A new Gene instance of the same type and with the same setup as this concrete Gene.</DL>
</DD>
</DL>
<HR>

<A NAME="cloneAllele()"><!-- --></A><H3>
cloneAllele</H3>
<PRE>
public <A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A> <B>cloneAllele</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#cloneAllele()">cloneAllele</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>clone<DT><B>See Also:</B><DD><A HREF="../../../org/jgapcustomised/Allele.html#cloneAllele()"><CODE>Allele.cloneAllele()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setValue(java.lang.Integer)"><!-- --></A><H3>
setValue</H3>
<PRE>
public void <B>setValue</B>(java.lang.Integer&nbsp;a_newValue)</PRE>
<DL>
<DD>Sets the value (allele) of this Gene to the new given value. This class expects the value to be an Integer
 instance. If the value is above or below the upper or lower bounds, it will be mappped to within the allowable
 range.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_newValue</CODE> - the new value of this Gene instance.</DL>
</DD>
</DL>
<HR>

<A NAME="getValue()"><!-- --></A><H3>
getValue</H3>
<PRE>
public java.lang.Integer <B>getValue</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>value</DL>
</DD>
</DL>
<HR>

<A NAME="getPersistentRepresentation()"><!-- --></A><H3>
getPersistentRepresentation</H3>
<PRE>
public java.lang.String <B>getPersistentRepresentation</B>()
                                             throws java.lang.UnsupportedOperationException</PRE>
<DL>
<DD>Retrieves a string representation of this Gene that includes any
 
 information required to reconstruct it at a later time, such as its
 
 value and internal state. This string will be used to represent this
 
 Gene in XML persistence. This is an optional method but, if not
 
 implemented, XML persistence and possibly other features will not be
 
 available. An UnsupportedOperationException should be thrown if no
 
 implementation is provided.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>A string representation of this Gene's current state.
<DT><B>Throws:</B>
<DD><CODE>java.lang.UnsupportedOperationException</CODE> - to indicate that no implementation
 
             is provided for this method.</DL>
</DD>
</DL>
<HR>

<A NAME="setValueFromPersistentRepresentation(java.lang.String)"><!-- --></A><H3>
setValueFromPersistentRepresentation</H3>
<PRE>
public void <B>setValueFromPersistentRepresentation</B>(java.lang.String&nbsp;a_representation)
                                          throws <A HREF="../../../org/jgapcustomised/UnsupportedRepresentationException.html" title="class in org.jgapcustomised">UnsupportedRepresentationException</A></PRE>
<DL>
<DD>Sets the value and internal state of this Gene from the string
 
 representation returned by a previous invocation of the
 
 getPersistentRepresentation() method. This is an optional method but,
 
 if not implemented, XML persistence and possibly other features will not
 
 be available. An UnsupportedOperationException should be thrown if no
 
 implementation is provided.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_representation</CODE> - the string representation retrieved from a
 
            prior call to the getPersistentRepresentation()
 
            method.
<DT><B>Throws:</B>
<DD><CODE>java.lang.UnsupportedOperationException</CODE> - to indicate that no implementation
 
             is provided for this method.
<DD><CODE><A HREF="../../../org/jgapcustomised/UnsupportedRepresentationException.html" title="class in org.jgapcustomised">UnsupportedRepresentationException</A></CODE> - if this Gene implementation
 
             does not support the given string representation.</DL>
</DD>
</DL>
<HR>

<A NAME="intValue()"><!-- --></A><H3>
intValue</H3>
<PRE>
public int <B>intValue</B>()</PRE>
<DL>
<DD>Retrieves the int value of this Gene, which may be more convenient in
 
 some cases than the more general getAllele() method.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the int value of this Gene.</DL>
</DD>
</DL>
<HR>

<A NAME="setToRandomValue(java.util.Random, boolean)"><!-- --></A><H3>
setToRandomValue</H3>
<PRE>
public void <B>setToRandomValue</B>(java.util.Random&nbsp;a_numberGenerator,
                             boolean&nbsp;onlyPerturbFromCurrentValue)</PRE>
<DL>
<DD>Sets the value (allele) of this Gene to a random Integer value between
 
 the lower and upper bounds (if any) of this Gene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#setToRandomValue(java.util.Random, boolean)">setToRandomValue</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_numberGenerator</CODE> - The random number generator that should be
 
            used to create any random values. It's important
 
            to use this generator to maintain the user's
 
            flexibility to configure the genetic engine
 
            to use the random number generator of their
 
            choice.<DD><CODE>onlyPerturbFromCurrentValue</CODE> - If true then the value should only be perturbed some small amount from the
            current value. If false then the value should be set to a completely new value irrespective of the
            current value.</DL>
</DD>
</DL>
<HR>

<A NAME="hashCode()"><!-- --></A><H3>
hashCode</H3>
<PRE>
public int <B>hashCode</B>()</PRE>
<DL>
<DD>Retrieves the hash code value for this IntegerGene.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#hashCode()">hashCode</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>this IntegerGene's hash code.<DT><B>See Also:</B><DD><CODE>Object.hashCode()</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD>Retrieves a string representation of this IntegerGene's value that
 
 may be useful for display purposes.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#toString()">toString</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a string representation of this IntegerGene's value.<DT><B>See Also:</B><DD><CODE>Object.toString()</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="cleanup()"><!-- --></A><H3>
cleanup</H3>
<PRE>
public void <B>cleanup</B>()</PRE>
<DL>
<DD>Executed by the genetic engine when this Gene instance is no
 
 longer needed and should perform any necessary resource cleanup.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="mapValueToWithinBounds()"><!-- --></A><H3>
mapValueToWithinBounds</H3>
<PRE>
protected void <B>mapValueToWithinBounds</B>()</PRE>
<DL>
<DD>Maps the value of this IntegerGene to within the bounds specified by
 
 the m_upperBounds and m_lowerBounds instance variables. The value's
 
 relative position within the integer range will be preserved within the
 
 bounds range (in other words, if the value is about halfway between the
 
 integer max and min, then the resulting value will be about halfway
 
 between the upper bounds and lower bounds). If the value is null or
 
 is already within the bounds, it will be left unchanged.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="calculateBoundsUnitsToIntegerUnitsRatio()"><!-- --></A><H3>
calculateBoundsUnitsToIntegerUnitsRatio</H3>
<PRE>
protected void <B>calculateBoundsUnitsToIntegerUnitsRatio</B>()</PRE>
<DL>
<DD>Calculates and sets the m_boundsUnitsToIntegerUnits field based
 
 on the current lower and upper bounds of this IntegerGene. For example,
 
 if the integer range is -2 billion to +2 billion and the bounds range
 
 is -1 billion to +1 billion, then each unit in the bounds range would
 
 map to 2 units in the integer range. The m_boundsUnitsToIntegerUnits
 
 field would therefore be 2. This mapping unit is used to map illegal
 
 allele values that are outside of the bounds to legal allele values that
 
 are within the bounds.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/IntegerAllele.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/jgapcustomised/impl/DefaultConfiguration.html" title="class in org.jgapcustomised.impl"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../org/jgapcustomised/impl/Pool.html" title="class in org.jgapcustomised.impl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/jgapcustomised/impl/IntegerAllele.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="IntegerAllele.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
