<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_27) on Fri Mar 22 11:50:03 EST 2013 -->
<TITLE>
NeuronAllele
</TITLE>

<META NAME="date" CONTENT="2013-03-22">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="NeuronAllele";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/NeuronAllele.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../com/anji/neat/NeatTargetFitnessFunction.html" title="class in com.anji.neat"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../com/anji/neat/NeuronGene.html" title="class in com.anji.neat"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?com/anji/neat/NeuronAllele.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="NeuronAllele.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
com.anji.neat</FONT>
<BR>
Class NeuronAllele</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">org.jgapcustomised.Allele</A>
      <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><B>com.anji.neat.NeuronAllele</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.lang.Comparable</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>NeuronAllele</B><DT>extends <A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></DL>
</PRE>

<P>
Gene corresponding to NEAT node gene according to <a
 href="http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf"> Evolving Neural Networks through Augmenting
 Topologies </a>
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD>Philip Tucker</DD>
</DL>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../com/anji/neat/NeuronAllele.html#NeuronAllele(com.anji.neat.NeuronGene)">NeuronAllele</A></B>(<A HREF="../../../com/anji/neat/NeuronGene.html" title="class in com.anji.neat">NeuronGene</A>&nbsp;aNeuronGene)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../com/anji/neat/NeuronAllele.html#cloneAllele()">cloneAllele</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../com/anji/neat/NeuronAllele.html#distance(org.jgapcustomised.Allele)">distance</A></B>(<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A>&nbsp;aTarget)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../com/anji/neat/NeuronAllele.html#getActivationType()">getActivationType</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/neat/NeuronType.html" title="class in com.anji.neat">NeuronType</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../com/anji/neat/NeuronAllele.html#getType()">getType</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../com/anji/neat/NeuronAllele.html#isType(com.anji.neat.NeuronType)">isType</A></B>(<A HREF="../../../com/anji/neat/NeuronType.html" title="class in com.anji.neat">NeuronType</A>&nbsp;aType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../com/anji/neat/NeuronAllele.html#setToRandomValue(java.util.Random, boolean)">setToRandomValue</A></B>(java.util.Random&nbsp;a_numberGenerator,
                 boolean&nbsp;onlyPerturbFromCurrentValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value of this Gene to a random legal value for the implementation.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.jgapcustomised.Allele"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class org.jgapcustomised.<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#compareTo(java.lang.Object)">compareTo</A>, <A HREF="../../../org/jgapcustomised/Allele.html#equals(java.lang.Object)">equals</A>, <A HREF="../../../org/jgapcustomised/Allele.html#getGene()">getGene</A>, <A HREF="../../../org/jgapcustomised/Allele.html#getGenes(java.util.Collection)">getGenes</A>, <A HREF="../../../org/jgapcustomised/Allele.html#getInnovationId()">getInnovationId</A>, <A HREF="../../../org/jgapcustomised/Allele.html#hashCode()">hashCode</A>, <A HREF="../../../org/jgapcustomised/Allele.html#toString()">toString</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, finalize, getClass, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="NeuronAllele(com.anji.neat.NeuronGene)"><!-- --></A><H3>
NeuronAllele</H3>
<PRE>
public <B>NeuronAllele</B>(<A HREF="../../../com/anji/neat/NeuronGene.html" title="class in com.anji.neat">NeuronGene</A>&nbsp;aNeuronGene)</PRE>
<DL>
<DL>
<DT><B>Parameters:</B><DD><CODE>aNeuronGene</CODE> - </DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="cloneAllele()"><!-- --></A><H3>
cloneAllele</H3>
<PRE>
public <A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A> <B>cloneAllele</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#cloneAllele()">cloneAllele</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>Gene clone of this object<DT><B>See Also:</B><DD><A HREF="../../../org/jgapcustomised/Allele.html#cloneAllele()"><CODE>Allele.cloneAllele()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setToRandomValue(java.util.Random, boolean)"><!-- --></A><H3>
setToRandomValue</H3>
<PRE>
public void <B>setToRandomValue</B>(java.util.Random&nbsp;a_numberGenerator,
                             boolean&nbsp;onlyPerturbFromCurrentValue)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../../org/jgapcustomised/Allele.html#setToRandomValue(java.util.Random, boolean)">Allele</A></CODE></B></DD>
<DD>Sets the value of this Gene to a random legal value for the implementation. This method exists for the benefit of
 mutation and other operations that simply desire to randomize the value of a gene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#setToRandomValue(java.util.Random, boolean)">setToRandomValue</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_numberGenerator</CODE> - The random number generator that should be used to create any random values. It's
            important to use this generator to maintain the user's flexibility to configure the genetic engine to
            use the random number generator of their choice.<DD><CODE>onlyPerturbFromCurrentValue</CODE> - If true then the value should only be perturbed some small amount from the
            current value. If false then the value should be set to a completely new value irrespective of the
            current value.<DT><B>See Also:</B><DD><A HREF="../../../org/jgapcustomised/Allele.html#setToRandomValue(java.util.Random, boolean)"><CODE>Allele.setToRandomValue(Random,boolean)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="distance(org.jgapcustomised.Allele)"><!-- --></A><H3>
distance</H3>
<PRE>
public double <B>distance</B>(<A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A>&nbsp;aTarget)</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgapcustomised/Allele.html#distance(org.jgapcustomised.Allele)">distance</A></CODE> in class <CODE><A HREF="../../../org/jgapcustomised/Allele.html" title="class in org.jgapcustomised">Allele</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>aTarget</CODE> - should be <code>NeuronAllele</code> with same gene TODO - activation type and slope
<DT><B>Returns:</B><DD>positive distance between genes, where a value closer to 0 represents more similar genes; used in
         computing distance between chromosomes, which in turn is used to compute speciation compatibility.<DT><B>See Also:</B><DD><A HREF="../../../org/jgapcustomised/Allele.html#distance(org.jgapcustomised.Allele)"><CODE>Allele.distance(org.jgapcustomised.Allele)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getType()"><!-- --></A><H3>
getType</H3>
<PRE>
public <A HREF="../../../com/anji/neat/NeuronType.html" title="class in com.anji.neat">NeuronType</A> <B>getType</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>neuron type<DT><B>See Also:</B><DD><A HREF="../../../com/anji/neat/NeuronGene.html#getType()"><CODE>NeuronGene.getType()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="isType(com.anji.neat.NeuronType)"><!-- --></A><H3>
isType</H3>
<PRE>
public boolean <B>isType</B>(<A HREF="../../../com/anji/neat/NeuronType.html" title="class in com.anji.neat">NeuronType</A>&nbsp;aType)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>aType</CODE> - 
<DT><B>Returns:</B><DD>true if <code>aType</code> matches<DT><B>See Also:</B><DD><A HREF="../../../com/anji/neat/NeuronGene.html#isType(com.anji.neat.NeuronType)"><CODE>NeuronGene.isType(NeuronType)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getActivationType()"><!-- --></A><H3>
getActivationType</H3>
<PRE>
public java.lang.String <B>getActivationType</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>activation type<DT><B>See Also:</B><DD><A HREF="../../../com/anji/neat/NeuronGene.html#getActivationType()"><CODE>NeuronGene.getActivationType()</CODE></A></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/NeuronAllele.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../com/anji/neat/NeatTargetFitnessFunction.html" title="class in com.anji.neat"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../com/anji/neat/NeuronGene.html" title="class in com.anji.neat"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?com/anji/neat/NeuronAllele.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="NeuronAllele.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
